### PR TITLE
update reconcileAliases to use opslevel-go implementation

### PR DESCRIPTION
**Pull request by davidbloss**
## Issues

[Combine reconcileAliases](https://github.com/OpsLevel/team-platform/issues/363)

## Changelog

Add `Config warning` when creating a resource with `aliases` where slug is missing.
Use new ReconcileAliases function from opslevel-go

- [X] List your changes here
- [x] Make a `changie` entry

### With this Terraform config:
```tf
resource "opslevel_infrastructure" "example" {
  aliases = []
  schema  = "Database"
  owner   = opslevel_team.example.id
  provider_data = {
    account = "dev2"
  }
  data = jsonencode({
    name = "my-database"
  })
}

resource "opslevel_service" "example" {
  name    = "foo25"
  aliases = []

  description = "The test service"
  framework   = "something fancy"
  language    = "elixir"

  lifecycle_alias = "beta"
  tier_alias      = "tier_2"
  owner           = opslevel_team.example.id

  api_document_path             = "/swagger.json"
  preferred_api_document_source = "PULL" //or "PUSH"
}

resource "opslevel_team" "example" {
  name             = "foo4"
  aliases          = []
  responsibilities = "Responsible for foo frontend and backend"

  member {
    email = "david+pat@opslevel.com"
    role  = "contributor"
  }
  member {
    email = "taimoor+pat@opslevel.com"
    role  = "contributor"
  }
  member {
    email = "kyle+pat@opslevel.com"
    role  = "manager"
  }
}
```

### Create resources with `aliases = []`, `terraform apply`
```tf
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # opslevel_infrastructure.example will be created
  + resource "opslevel_infrastructure" "example" {
      + aliases       = []
      + data          = jsonencode(
            {
              + name = "my-database"
            }
        )
      + id            = (known after apply)
      + owner         = (known after apply)
      + provider_data = {
          + account = "dev2"
        }
      + schema        = "Database"
    }

  # opslevel_service.example will be created
  + resource "opslevel_service" "example" {
      + aliases                       = []
      + api_document_path             = "/swagger.json"
      + description                   = "The test service"
      + framework                     = "something fancy"
      + id                            = (known after apply)
      + language                      = "elixir"
      + lifecycle_alias               = "beta"
      + name                          = "foo25"
      + owner                         = (known after apply)
      + preferred_api_document_source = "PULL"
      + tier_alias                    = "tier_2"
    }

  # opslevel_team.example will be created
  + resource "opslevel_team" "example" {
      + aliases          = []
      + id               = (known after apply)
      + name             = "foo4"
      + responsibilities = "Responsible for foo frontend and backend"

      + member {
          + email = "david+pat@opslevel.com"
          + role  = "contributor"
        }
      + member {
          + email = "kyle+pat@opslevel.com"
          + role  = "manager"
        }
      + member {
          + email = "taimoor+pat@opslevel.com"
          + role  = "contributor"
        }
    }

Plan: 3 to add, 0 to change, 0 to destroy.
opslevel_team.example: Creating...
opslevel_team.example: Creation complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUxOA]
opslevel_infrastructure.example: Creating...
opslevel_service.example: Creating...
opslevel_infrastructure.example: Creation complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTUwNTg]
opslevel_service.example: Creation complete after 2s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3MjI]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```
### Destroy resources, `terraform destroy` ✅  
-----

### With config `aliases` set, but without value in `name`
```tf
resource "opslevel_team" "example" {
  name             = "foo4"
  aliases          = ["newnewnew"]   # <-- slug is missing here
  responsibilities = "Responsible for foo frontend and backend"

  member {
    email = "david+pat@opslevel.com"
    role  = "contributor"
  }
  member {
    email = "taimoor+pat@opslevel.com"
    role  = "contributor"
  }
  member {
    email = "kyle+pat@opslevel.com"
    role  = "manager"
  }
}
```

### Run `terraform apply`, note missing alias `foo4_11`
```tf
# ... truncated output
Plan: 1 to add, 0 to change, 0 to destroy.
opslevel_team.example: Creating...
╷
│ Warning: opslevel client error
│ 
│   with opslevel_team.example,
│   on main.tf line 29, in resource "opslevel_team" "example":
│   29: resource "opslevel_team" "example" {
│ 
│ warning while reconciling team aliases: '[newnewnew]'
│ OpsLevel API Errors:
│       - 'alias' slug is locked, it cannot be deleted
│ 
│ OpsLevel API Errors:
│       - 'alias' Alias 'newnewnew' has already been taken
│ 
╵
╷
│ Warning: Config warning
│ 
│   with opslevel_team.example,
│   on main.tf line 29, in resource "opslevel_team" "example":
│   29: resource "opslevel_team" "example" {
│ 
│ On create, OpsLevel API creates a new alias for teams. If this causes issues, create team with empty 'aliases'. Then update team with 'aliases'
╵
╷
│ Error: Config error
│ 
│   with opslevel_team.example,
│   on main.tf line 29, in resource "opslevel_team" "example":
│   29: resource "opslevel_team" "example" {
│ 
│ default aliases from API need to be added to config: [foo4_11 newnewnew]
```

### Add missing alias, `terraform apply` again (still fails - needs alias `foo4_12` now)
```tf
# ... truncated output
│ Warning: Config warning
│ 
│   with opslevel_team.example,
│   on main.tf line 29, in resource "opslevel_team" "example":
│   29: resource "opslevel_team" "example" {
│ 
│ On create, OpsLevel API creates a new alias for teams. If this causes issues, create team with empty 'aliases'. Then update team with 'aliases'
╵
╷
│ Error: Config error
│ 
│   with opslevel_team.example,
│   on main.tf line 29, in resource "opslevel_team" "example":
│   29: resource "opslevel_team" "example" {
│ 
│ default aliases from API need to be added to config: [foo4_12 foo4_11 newnewnew]
```

----
### With this config, note each resource has `aliases = []`
```tf
resource "opslevel_infrastructure" "example" {
  aliases = []
  schema  = "Database"
  owner   = opslevel_team.example.id
  provider_data = {
    account = "dev2"
  }
  data = jsonencode({
    name = "my-database"
  })
}

resource "opslevel_service" "example" {
  name    = "foo25"
  aliases = []

  description = "The test service"
  framework   = "something fancy"
  language    = "elixir"

  lifecycle_alias = "beta"
  tier_alias      = "tier_2"
  owner           = opslevel_team.example.id

  api_document_path             = "/swagger.json"
  preferred_api_document_source = "PULL" //or "PUSH"
}

resource "opslevel_team" "example" {
  name             = "foo4"
  aliases          = []
  responsibilities = "Responsible for foo frontend and backend"

  member {
    email = "david+pat@opslevel.com"
    role  = "contributor"
  }
  member {
    email = "taimoor+pat@opslevel.com"
    role  = "contributor"
  }
  member {
    email = "kyle+pat@opslevel.com"
    role  = "manager"
  }
}
```

### Create resources, `terraform apply`
```tf
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # opslevel_infrastructure.example will be created
  + resource "opslevel_infrastructure" "example" {
      + aliases       = []
      + data          = jsonencode(
            {
              + name = "my-database"
            }
        )
      + id            = (known after apply)
      + owner         = (known after apply)
      + provider_data = {
          + account = "dev2"
        }
      + schema        = "Database"
    }

  # opslevel_service.example will be created
  + resource "opslevel_service" "example" {
      + aliases                       = []
      + api_document_path             = "/swagger.json"
      + description                   = "The test service"
      + framework                     = "something fancy"
      + id                            = (known after apply)
      + language                      = "elixir"
      + lifecycle_alias               = "beta"
      + name                          = "foo25"
      + owner                         = (known after apply)
      + preferred_api_document_source = "PULL"
      + tier_alias                    = "tier_2"
    }

  # opslevel_team.example will be created
  + resource "opslevel_team" "example" {
      + aliases          = []
      + id               = (known after apply)
      + name             = "foo4"
      + responsibilities = "Responsible for foo frontend and backend"

      + member {
          + email = "david+pat@opslevel.com"
          + role  = "contributor"
        }
      + member {
          + email = "kyle+pat@opslevel.com"
          + role  = "manager"
        }
      + member {
          + email = "taimoor+pat@opslevel.com"
          + role  = "contributor"
        }
    }

Plan: 3 to add, 0 to change, 0 to destroy.
opslevel_team.example: Creating...
opslevel_team.example: Creation complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA]
opslevel_infrastructure.example: Creating...
opslevel_service.example: Creating...
opslevel_infrastructure.example: Creation complete after 0s [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc]
opslevel_service.example: Creation complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

### Update config, aliases are set but slug is missing
```tf
resource "opslevel_infrastructure" "example" {
  aliases = ["infratest"]
  schema  = "Database"
  owner   = opslevel_team.example.id
  provider_data = {
    account = "dev2"
  }
  data = jsonencode({
    name = "my-database"
  })
}

resource "opslevel_service" "example" {
  name    = "foo25"
  aliases = ["servicetest"]

  description = "The test service"
  framework   = "something fancy"
  language    = "elixir"

  lifecycle_alias = "beta"
  tier_alias      = "tier_2"
  owner           = opslevel_team.example.id

  api_document_path             = "/swagger.json"
  preferred_api_document_source = "PULL" //or "PUSH"
}

resource "opslevel_team" "example" {
  name             = "foo4"
  aliases          = ["teamtest"]
  responsibilities = "Responsible for foo frontend and backend"

  member {
    email = "david+pat@opslevel.com"
    role  = "contributor"
  }
  member {
    email = "taimoor+pat@opslevel.com"
    role  = "contributor"
  }
  member {
    email = "kyle+pat@opslevel.com"
    role  = "manager"
  }
}
```

### Attempt to updated resources, needs slugs raised by `Config error`
```tf
# ...truncted output
│ Error: Config error
│ 
│   with opslevel_team.example,
│   on main.tf line 29, in resource "opslevel_team" "example":
│   29: resource "opslevel_team" "example" {
│ 
│ default aliases from API need to be added to config: [foo4_13]

### Add `foo4_13` to `aliases` then `terraform apply`. Infra and Team are created, service has same alias issue
```tf
opslevel_team.example: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA]
opslevel_infrastructure.example: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc]
opslevel_service.example: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_infrastructure.example will be updated in-place
  ~ resource "opslevel_infrastructure" "example" {
      ~ aliases       = [
          + "infratest",
        ]
        id            = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc"
        # (4 unchanged attributes hidden)
    }

  # opslevel_service.example will be updated in-place
  ~ resource "opslevel_service" "example" {
      ~ aliases                       = [
          + "servicetest",
        ]
        id                            = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA"
        name                          = "foo25"
        # (8 unchanged attributes hidden)
    }

  # opslevel_team.example will be updated in-place
  ~ resource "opslevel_team" "example" {
      + aliases          = [
          + "foo4_13",
          + "teamtest",
        ]
        id               = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA"
        name             = "foo4"
        # (1 unchanged attribute hidden)

      + member {
          + email = "david+pat@opslevel.com"
          + role  = "contributor"
        }
      + member {
          + email = "kyle+pat@opslevel.com"
          + role  = "manager"
        }
      + member {
          + email = "taimoor+pat@opslevel.com"
          + role  = "contributor"
        }
    }

Plan: 0 to add, 3 to change, 0 to destroy.
opslevel_team.example: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA]
opslevel_team.example: Modifications complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA]
opslevel_infrastructure.example: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc]
opslevel_service.example: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA]
opslevel_infrastructure.example: Modifications complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc]
╷
│ Warning: opslevel client error
│ 
│   with opslevel_service.example,
│   on main.tf line 13, in resource "opslevel_service" "example":
│   13: resource "opslevel_service" "example" {
│ 
│ Unable to reconcile service aliases: '[servicetest]'
│ OpsLevel API Errors:
│       - 'alias' slug is locked, it cannot be deleted
│ 
╵
╷
│ Error: Config error
│ 
│   with opslevel_service.example,
│   on main.tf line 13, in resource "opslevel_service" "example":
│   13: resource "opslevel_service" "example" {
│ 
│ default aliases from API need to be added to config: [foo25]
```

### Fix service alias, `aliases = ["servicetest", "foo25"]`, `terraform apply`
```tf
opslevel_team.example: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA]
opslevel_infrastructure.example: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc]
opslevel_service.example: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_service.example will be updated in-place
  ~ resource "opslevel_service" "example" {
      ~ aliases                       = [
          + "foo25",
          + "servicetest",
        ]
        id                            = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA"
        name                          = "foo25"
        # (8 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
opslevel_service.example: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA]
opslevel_service.example: Modifications complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

### Unset aliases (`aliases = null`) for all resources then `terraform apply`
```tf
opslevel_team.example: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA]
opslevel_service.example: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA]
opslevel_infrastructure.example: Refreshing state... [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # opslevel_infrastructure.example will be updated in-place
  ~ resource "opslevel_infrastructure" "example" {
      - aliases       = [
          - "infratest",
        ] -> null
        id            = "Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc"
        # (4 unchanged attributes hidden)
    }

  # opslevel_service.example will be updated in-place
  ~ resource "opslevel_service" "example" {
      - aliases                       = [
          - "foo25",
          - "servicetest",
        ] -> null
        id                            = "Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA"
        name                          = "foo25"
        # (8 unchanged attributes hidden)
    }

  # opslevel_team.example will be updated in-place
  ~ resource "opslevel_team" "example" {
      - aliases          = [
          - "foo4_13",
          - "teamtest",
        ] -> null
        id               = "Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA"
        name             = "foo4"
        # (1 unchanged attribute hidden)

        # (3 unchanged blocks hidden)
    }

Plan: 0 to add, 3 to change, 0 to destroy.
opslevel_team.example: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA]
opslevel_team.example: Modifications complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvVGVhbS8xOTUzOA]
opslevel_infrastructure.example: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc]
opslevel_service.example: Modifying... [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA]
opslevel_infrastructure.example: Modifications complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvRW50aXR5T2JqZWN0LzI3NTY2NTc]
opslevel_service.example: Modifications complete after 1s [id=Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMjU3NDA]
╷
│ Warning: opslevel client error
│ 
│   with opslevel_team.example,
│   on main.tf line 29, in resource "opslevel_team" "example":
│   29: resource "opslevel_team" "example" {
│ 
│ warning while reconciling team aliases: '[]'
│ OpsLevel API Errors:
│       - 'alias' slug is locked, it cannot be deleted
│ 
╵

Apply complete! Resources: 0 added, 3 changed, 0 destroyed.
```

### `terraform destroy` ✅

### CODE REVIEW

File: examples/resources/opslevel_service/resource.tf [hunk 1]
```diff
@@ -15,7 +15,7 @@ data "opslevel_tier" "tier3" {
 resource "opslevel_team" "foo" {
   name             = "foo"
   responsibilities = "Responsible for foo frontend and backend"
-  aliases          = ["bar", "baz"]
+  aliases          = ["foo", "bar", "baz"] # NOTE: if set, value of "name" must be included
```
Comment (rocktavious) [inline examples/resources/opslevel_service/resource.tf:18]: not entirely true more like

```suggestion
  aliases          = ["foo", "bar", "baz"] # NOTE: if set, slugified value of "name" must be included
```

File: opslevel/common.go [hunk 2]
```diff
@@ -111,6 +113,55 @@ func FilterAttrs(validFieldNames []string) map[string]schema.Attribute {
 	return filterAttrs
 }
 
+// return strings not in both slices
+func diffBetweenStringSlices(sliceOne, sliceTwo []string) []string {
```
Comment (rocktavious) [inline opslevel/common.go:117]: surprising the `slices` package doesn't have something like this.

File: opslevel/import_utils.go [hunk 3]
```diff
@@ -53,6 +53,13 @@ func getTagsFromResource(client *opslevel.Client, resource opslevel.TaggableReso
 	return tags, diags
 }
 
+// hasTagFormat returns true if the given tag is formatted as '<key>:<value>'
+func hasTagFormat(tag string) bool {
```
Comment (rocktavious) [inline opslevel/import_utils.go:57]: Why is `isTagValid` different from this?  is that function miss named?

Comment (davidbloss) [reply opslevel/import_utils.go:57]: `isImportTagValid` would be a better name. I can add that in

File: opslevel/resource_opslevel_team.go [hunk 4]
```diff
@@ -149,30 +143,33 @@ func (teamResource *TeamResource) Create(ctx context.Context, req resource.Creat
...
+		}
+		if err = team.ReconcileAliases(teamResource.client, aliases); err != nil {
+			resp.Diagnostics.AddWarning("opslevel client error", fmt.Sprintf("warning while reconciling team aliases: '%s'\n%s", aliases, err))
+			resp.Diagnostics.AddWarning("Config warning", "On create, OpsLevel API creates a new alias for teams. If this causes issues, create team with empty 'aliases'. Then update team with 'aliases'")
+		}
 	}
 
 	createdTeamResourceModel, diags := NewTeamResourceModel(ctx, *team, planModel)
 	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
```
Comment (rocktavious) [inline opslevel/resource_opslevel_team.go:170]: this is logically different then `if diags.HasError()` are we sure we want to bail here if there has been any error added up until this point and not just if `NewTeamResourceModel` had errors?

Comment (davidbloss) [reply opslevel/resource_opslevel_team.go:170]: 🤔 In this function, every `resp.Diagnostics.AddError(...)` is followed by a `return` and there is no `resp.Diagnostics.Append(diags...)` before the line above. Each `if diags.HasError() { ... }` contains a `return`. I'm not seeing the case where an error was added to `resp.Diagnostics` but not returned

Comment (rocktavious) [reply opslevel/resource_opslevel_team.go:170]: then why `resp.Diagnostics.Append(diags...)` ?

Comment (rocktavious) [reply opslevel/resource_opslevel_team.go:170]: oh because warnings 🤦

File: examples/resources/opslevel_service/resource.tf (hunk 1 above)
Comment (rocktavious) [reply examples/resources/opslevel_service/resource.tf]: Theres a few more too

File: opslevel/common.go [hunk 5]
```diff
@@ -111,6 +113,55 @@ func FilterAttrs(validFieldNames []string) map[string]schema.Attribute {
...
+
+	// check if config has 'aliases' set, but is missing needed alias "slugs" from API
+	aliasesFromModel, diags := SetValueToStringSlice(ctx, modelAliases)
+	if diags != nil && diags.HasError() {
+		return aliases, diags
+	}
+	aliasesNeededInConfig := diffBetweenStringSlices(resourceAliases, aliasesFromModel)
+	if len(aliasesNeededInConfig) > 0 {
+		// setting aliases here blocks an erroneous error. The "Config error" is what matters
+		diags.AddError("Config error", fmt.Sprintf(`default aliases from API need to be added to config: %s`, aliasesNeededInConfig))
```
Comment (rocktavious) [inline opslevel/common.go:157]: VERY NICE - this will greatly help customers!!!

**Review by rocktavious — APPROVED** [review]
🥵

**Pull request merged**